---
title: "IEO Project Analysis"
author:
- name: "Radostina Kisleva, Emma Juan Salazar, Tamara-Dora Veres"
  affiliation: Universitat Pompeu Fabra
  email: radostina.kisleva01@estudiant.upf.edu,  emma.juan01@estudiant.upf.edu,  tamaradora.veres01@estudiant.upf.edu
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
    fig_captions: yes
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{IEOprojectAnalysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!--
In this first chunk of code, which will not be shown in the resulting
document (echo=FALSE) sets up global processing options, such as whether
a comment character should appear before code results (set to the null
character string in this case), collapse source and output blocks into
a single code block (collapse=TRUE), align figures to the center
(fig.align="center") or cache the results to speed up vignette building
(cache=FALSE thus disabled in this case). A full description of possible
options can be found at http://yihui.name/knitr/options
--->

```{r setup, echo=FALSE, cache=FALSE}
library(knitr) ## kable()
library(kableExtra) ## kable_styling(), save_kable()
library(here) ## here()
library(usethis) ## use_directory()

knitr::opts_chunk$set(
  collapse=TRUE,
  comment="",
  fig.align="center",
  fig.wide=TRUE,
  cache=FALSE
)

## this option avoid use_directory() being verbose
options(usethis.quiet=TRUE)

## create these paths at build time if they do not exist
use_directory(file.path("doc"))
use_directory(file.path("inst", "doc"))

## fetch the package root directory
path2pkg <- here()

```

# Introduction

Lung cancer is the leading cause of cancer-related death in the United
States. The molecular events preceding the onset of disease are poorly
understood, and no effective tools exist to identify smokers with
premalignant lesions (PMLs) that will progress to invasive cancer.
Previous research has identified molecular changes in the smoke-exposed
airways. However, the focus of the paper "Detecting the Presence and
Progression of Premalignant Lung Lesions via Airway Gene
Expression" @Beane17 shifts the focus to the first stages of the disease progress.
They utilize samples from the same
airway field of injury to study PMLs and their potential in preventing
lung cancer.

The researchers' goal is to profile samples extracted from normal
appearing bronchial mucosa by mRNA-Seq, and attempt to find
differentially expressed gene sets and pathways between subjects with
PMLs and without PMLs. From the original 82 samples, 50 subjects had
PML, 25 didn't, and 7 were excluded in the end.

The resulting raw RNA-seq data has been deposited at the Gene
Expression Omnibus (GEO), where they are publicly available under
accession
[GSE79209](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE79209).
Here, we show a first exploratory analysis of the corresponding RNA-seq
gene expression profiles generated as a table of counts using the DEE2
(<https://dee2.io>) pipeline by @ziemann19, and further packaged into a
[SummarizedExperiment](https://bioconductor.org/packages/SummarizedExperiment)
object with genes mapped to Entrez identifiers. This object also stores
the phenotypic information about the profiled samples that has been also
made available at GEO.

# Quality assessment

## Data import and cleaning

We start importing the raw table of counts.

<!--
The option 'message=FALSE' avoid dumping R messages such as "Loading required package: methods"
into the output of the report.
-->

```{r, message=FALSE}
library(SummarizedExperiment)

se <- readRDS(file.path(system.file("extdata",
                                    package="IEOproject"),
                        "GSE79209.rds"))
se

```

We have `r nrow(se)` genes by `r ncol(se)` samples. From the first row
and column names shown by the object, we can figure out that genes are
defined by [Entrez](https://www.ncbi.nlm.nih.gov/gene) [@maglott10]
identifiers and samples by Sequence Read Archive Run
([SRR](https://www.ncbi.nlm.nih.gov/books/NBK56913/#search.what_do_the_different_sra_accessi)) identifiers.

The row data in this object contains information about the profiled genes.

```{r}
head(rowData(se))
```
Among this information, the gene symbol and description are potentially
useful for interpreting results of, for instance, a differential
expression analysis. Let's explore now the column (phenotypic) data.

```{r}
dim(colData(se))
head(colData(se), n=3)
```

We have a total of `r ncol(colData(se))` phenotypic variables. The
second column `geo_accession` contains GEO Sample Accession Number
([GSM](https://www.ncbi.nlm.nih.gov/geo/info/overview.html)) identifers.
GSM identifiers define individual samples, understood in our context as
individual sources of RNA. We can check if we have any technical replicates
as follows:

```{r}
length(unique(se$geo_accession))
table(lengths(split(colnames(se), se$geo_accession)))
```

So, we have `r length(unique(se$geo_accession))` different individual
samples which matches the number of samples that we have `r ncol(se)`.
Therefore we don't have any technical replicates and we can proceed with
the next step.

To proceed further exploring this dataset, we are going to use the
[edgeR](https://bioconductor.org/packages/edgeR) package and build
a `DGEList` object, incorporating the gene metadata, which includes
the gene symbol.

```{r, message=FALSE}
library(edgeR)

dge <- DGEList(counts=assays(se)$counts, genes=rowData(se))
head(dge$samples, 3)
dim(dge)
```
Calculate $\log_2$ CPM units of expression and put them as an
additional assay element to ease their manipulation. This provides a
standarized and normalized representation of gene expression data.

```{r}
assays(se)$logCPM <- cpm(dge, log=TRUE)
assays(se)$logCPM[1:5, 1:5]
```

Let's explore now some of the phenotypic variables. After some visual inspection
of the variables, we find out that the variable `se$characteristics_ch1.4` contains
information about how many cigarettes a person has smoked in their lifetime.
We also have a variable which tells us whether the person has COPD 
(Chronic obstructive pulmonary disease) or not `se$characteristics_ch1.5`.

```{r}
table(se$characteristics_ch1.4)
table(se$characteristics_ch1.5)
```

Another variable of interest could be max histology `se$characteristics_ch1.6` which 
represents the different histology types (tissue characteristics).

```{r}
table(se$characteristics_ch1.6)
```

To facilitate handling this variables we are going to perform the following
modifications:

```{r}
se$copd_status <- gsub("copd status: ", "", se$characteristics_ch1.5)
se$copd_status <- factor(se$copd_status, levels = c("COPD", "No COPD"))

se$histology <- gsub("max histology: ", "", se$characteristics_ch1.6)
se$histology <- factor(se$histology, levels(se$histology) <- c("Hyperplasia", "Metaplasia", "Mild dysplasia", "Moderate dysplasia", "Normal", "Severe dysplasia"))
```

To faciliate working with the pack years variable will split it into four 
categorical groups - Low, Medium, High and Very high. 

To do that we will:  
1. Calculate quartiles for the data.  
2. Assign categories based on quartile ranges.  
3. Convert the values into their corresponding categories.  


```{r}
se$characteristics_ch1.4 <- gsub("pack years: ", "", se$characteristics_ch1.4)
class(se$characteristics_ch1.4)
se$characteristics_ch1.4 <- as.numeric(se$characteristics_ch1.4)

quartiles <- quantile(se$characteristics_ch1.4, probs = c(0.25, 0.5, 0.75))

```

Now that we have the quartiles, we can proceed to assign categories to the values based on these quartiles.

Values below the first quartile (Q1) will be categorized as "low."
Values between Q1 and the median (Q2) will be categorized as "medium."
Values between the median (Q2) and the third quartile (Q3) will be categorized as "high."
Values above the third quartile (Q3) will be categorized as "very high."


```{r}

se$pack_years <- cut(se$characteristics_ch1.4,
                          breaks = c(-Inf, quartiles[1], quartiles[2], quartiles[3], Inf),
                          labels = c("Low", "Medium", "High", "Very High"),
                          include.lowest = TRUE)

```

In Table \@ref(tab:pheno) below, we show the three variables of interest
(histology, COPD status and pack years) jointly to gather as much understanding as
possible on the underlying experimental design.

```{r pheno, echo=FALSE, message=FALSE}
tmpdf <- data.frame("Identifer"=colnames(se),
                    "Pack Years"=se$pack_years,
                    Histology=se$histology,
                    "COPD Status"=se$copd_status,
                    check.names=FALSE)
ktab <- kable(tmpdf, caption="Phenotypic variables.")
kable_styling(ktab, position="center")
```

## Sequencing depth

In the following step we will evaluate the sequencing depth in terms of the total number of read counts that are mapped to each sample's genome. The sequencing depth per sample, commonly referred to as library sizes, is displayed in Figure \@ref(fig:libsizes) below in increasing order. 
<!---
you can control the height and width in pixels of the figure with
'out.height' and 'out.width'. Figures are automatically numbered,
to refer to them in the main text you should use the notation shown
above as \@ref(fig:xxxx) with xxxx being the label in the code chunk
that also gives the filename of the figure. This name must be unique
--->

```{r libsizes, echo=FALSE, height=8, width=8, out.width="600px", fig.cap="Library sizes in increasing order."}
par(mar=c(7, 5, 2, 2))
ord <- order(dge$sample$lib.size/1e6)
ordmreads <- dge$sample$lib.size[ord]/1e6
names(ordmreads) <- colnames(se)[ord]
bp <- barplot(ordmreads, las=1, ylab="Millions of reads",
              xlab="", col=c("blue", "green", "red", "purple", "orange", "pink")[factor(se$histology[ord])], las=2, ylim = c(0, 50))
legend("topleft", c("Hyperplasia", "Metaplasia", "Mild dysplasia", "Moderate dysplasia", "Normal", "Severe dysplasia"), fill=c("blue", "green", "red", "purple", "orange", "pink"), inset=0.01, cex=0.85)

```

The plot displays a range of sequencing depths for the samples, each represented by an SRR identifier. The sequencing depth across our samples ranges from 20 to 50 million reads, with the majority falling between 30 to 40 million reads. This indicates a generally high level of coverage, which is beneficial for accurate genomic analysis. The plot also suggests that the highest number of reads corresponds to mild and moderate dysplasia, indicating early stages of cellular abnormalities that could potentially progress to cancer. The predominance of mild dysplasia in these samples could be of particular interest for early detection and intervention studies. There does not appear to be any evident bias in terms of sequencing depth across the conditions represented.



## Distribution of expression levels among samples

Figure \@ref(fig:distRawExp) below shows the distribution of expression
values per sample in logarithmic CPM units of expression.

<!---
the option echo=FALSE hides the R code. When plotting in general one
does not want to see the code. Options fig.height and fig.width control
height and width of the plot in inches while out.height and out.width
do it in the final output file; see http://yihui.name/knitr/options for
full details.
--->

```{r distRawExp, echo=FALSE, fig.height=5, fig.width=5, out.width="600px", fig.cap="Non-parametric density distribution of expression profiles per sample.", message=FALSE}
library(geneplotter)
par(mar=c(4, 5, 1, 1))
lst <- as.list(as.data.frame(assays(se)$logCPM))
multidensity(lst, xlab="log 2 CPM", legend=NULL,
             main="", las=1)
```
There are no substantial differences between the samples in the
distribution of expression values.

## Distribution of expression levels among genes

Let's calculate now the average expression per gene through all the
samples. Figure \@ref(fig:exprdist) shows the distribution of those
values across genes.

```{r exprdist, echo=FALSE, out.width="600px", fig.cap="Distribution of average expression level per gene."}
avgexp <- rowMeans(assays(se)$logCPM)
hist(avgexp, xlab="log2 CPM", main="", las=1)
```
The histogram displays a bimodal distribution, which means there are two distinct peaks. The large peak to the left is the most prominent feature of the histogram, showing that most genes in this dataset are expressed at low levels or possibly not at all, under the conditions of the experiment. On the right part, the peak which indicates that as gene expression levels increase, fewer genes are found at these higher expression levels.


## Filtering of lowly-expressed genes

In order to filter lowly-expressed genes, we calculate row means of expression data and then we apply logical masks. We filter out genes which do not meet certain criteria such as minimum expression levels and a sample threshold. This step is needed to refine the dataset and remove noise, ensuring the subsequent analysis is based on high-quality data. Figure \@ref(fig:lowexpgenes) shows the distribution of those
values across genes.

```{r}

mask_initial <- rowMeans(assays(se)$logCPM) > 1
se_initial_filt <- se[mask_initial, ]
dge_initial_filt <- dge[mask_initial, ]
dim(se_initial_filt)

cpmcutoff <- round(10/min(dge$samples$lib.size/1e6), digits=1)
cpmcutoff

if(any(is.na(se$histology))) {
  stop("Missing values found in the histology column.")
}

pml_status <- se$histology %in% c("Mild dysplasia", "Moderate dysplasia", "Severe dysplasia")
nsamplescutoff <- sum(pml_status)
nsamplescutoff

pml_indices <- which(pml_status)

cpm_pml <- cpm(dge_initial_filt)[, pml_indices]

mask_final <- rowSums(cpm_pml > cpmcutoff) >= nsamplescutoff
sum(mask_final)
se_final_filt <- se_initial_filt[mask_final, ]
dge_final_filt <- dge_initial_filt[mask_final, ]
dim(se_final_filt)
```

```{r lowexpgenes, echo=FALSE, out.width="600px", fig.cap="Distribution of lowly-expressed genes."}
par(mar=c(4, 5, 1, 1))
avgexp <- rowMeans(assays(se)$logCPM)
h <- hist(avgexp, xlab=expression("Expression level (" * log[2] * "CPM)"),
          main="", las=1, col="grey", cex.axis=1.2, cex.lab=1.5)
x <- cut(rowMeans(assays(se.filt)$logCPM), breaks=h$breaks)
lines(h$mids, table(x), type="h", lwd=10, lend=1, col="red")
legend("topright", c("All genes", "Filtered genes"), fill=c("grey", "red"))

```

We are left with `r nrow(se.filt)` genes.

## Normalization

We calculate now the normalization factors on the filtered expression
data set.

```{r}
dge.filt <- calcNormFactors(dge.filt)
```

Replace the raw log2 CPM units in the corresponding assay element of
the `SummarizedExperiment` object, by the normalized ones.

```{r}
assays(se.filt)$logCPM <- cpm(dge.filt, log=TRUE,
                              normalized.lib.sizes=TRUE)
```

## MA-plots

We examine now the MA-plots of the normalized expression profiles
in the figures below.

<!---
Here we make a MA-plot for each sample. The options 'fig.height'
and 'fig.width' control the relative image size in *inches*. The
final image size results from 'height'x'dpi' and 'width'x'dpi',
where 'dpi' is the image resolution in "dots per inch" (by default
dpi=72). To scale the image to a desired size use 'out.width' and
'out.height'. More information at http://yihui.name/knitr/options
--->
```{r maPlots, fig.height=18, fig.width=10, dpi=100, echo=FALSE, fig.cap="MA-plots of filtered and normalized expression values."}
par(mfrow=c(5, 4), mar=c(4, 5, 3, 1))
for (i in 1:ncol(se.filt)) {
  A <- rowMeans(assays(se.filt)$logCPM)
  M <- assays(se.filt)$logCPM[, i] - A
  smoothScatter(A, M, main=colnames(se.filt)[i], las=1)
  abline(h=0, col="blue", lwd=2)
  lo <- lowess(M ~ A)
  lines(lo$x, lo$y, col="red", lwd=2)
}
```

For genes above the horizontal line at M=0, expression is higher in the condition being tested against the reference, as for genes below, expression is lower.
The red line represents a loess fit to the data, which should ideally be around M=0 if there is no differential expression between conditions. Any systematic deviation from M=0 indicates potential bias or systematic variation in the data. In most of the plots, the red line remains close to M=0 across the range of A values, indicating no systematic bias in expression changes relative to gene abundance.
Some plots show a slight dip in the loess line at the lower end of the A scale, suggesting a potential bias for lowly expressed genes which could be potential candidates for differential expression. These points would be of interest for follow-up analyses.

## Experimental design and batch identification

To identify potential columns which could constitute unwanted variability, we looked at columns that contain information related to technical aspects, experimental conditions or processing details that could vary between different batches. Here are some columns which we thought could introduce unwanted variability: `submission_date`, `last_update_date`, `type`, `channel_count`, `extract_protocol_ch1`, `extract_protocol_ch1.1`, `data_processing`, `data_processing.1`, `data_processing.2`, `data_processing.3`, `data_processing.4`, `data_processing.5`, `data_processing.6`, `platform_id`, `instrument_model`, `library_selection`, `library_source`, `library_strategy`.

These variables have the same values across all samples which suggests that they are not contributing to the variability of the dataset. For example:
```{r}
table(se.filt$histology, se.filt$library_strategy)
```

There is perfect correlation between library strategy and histology. This means that differences between our samples will be due to biological differences rather than technical. 

Now let's examine the distribution between pack-years and histology.

```{r}
table(se.filt$pack_years, se.filt$histology)
```

We can see that not all combinations of pack-years and histology have
been sequenced. Some combinations have missing values, indicated by 0s. However, 
with only two missing combinations out of the several possible, the
impact on the overall analysis may be minimal and not critical for the analysis. 

We examine now how samples group together by hierarchical clustering
and multidimensional scaling, annotating pack-years and histology. We
calculate again log CPM values with a high prior count(3) to moderate extreme
fold-changes produced by low counts. The resulting dendrogram is shown in
Figure \@ref(fig:sampleClustering).

```{r sampleClustering, fig.height=8, fig.width=10, dpi=100, echo=FALSE, fig.cap="Figure S6: Hierarchical clustering of the samples. Labels correspond to pack-years and sample identifer, while colors indicate histology groups."}
logCPM <- cpm(dge.filt, log=TRUE, prior.count=3)
d <- as.dist(1-cor(logCPM, method="spearman"))
sampleClustering <- hclust(d)
batch <- as.integer(se.filt$histology)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- paste(se.filt$pack_years, colnames(se), sep="\n")
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
                               function(x, batch, labels) {
                                 if (is.leaf(x)) {
                                   attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x, "label")]))
                                   attr(x, "label") <- as.vector(labels[attr(x, "label")])
                                 }
                                 x
                               }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples",
     cex=1)
legend("topright", levels(se.filt$histology),
       fill=seq_len(nlevels(se.filt$histology)))

```

Samples are linked together based on their similarity, with the branches connecting those most similar first. As we move up the tree, the connections represent progressively less similar groups until all are joined in a single cluster. Samples do not really cluster together by histology type, which means histology type is not the primary factor driving the similarity or dissimilarity between samples.

In Figure \@ref(fig:mdsPlot) we show the
corresponding MDS plot.

```{r mdsPlot, fig.height=5, fig.width=8, dpi=100, echo=FALSE, fig.cap="Figure S7: Multidimensional scaling plot of the samples. Labels correspond to treatment and colors indicate sample group."}
outcome <- se.filt$pack_years
names(outcome) <- colnames(se.filt)
plotMDS(dge.filt, labels=outcome, col=batch)
legend("bottomleft", levels(se.filt$histology),
       fill=seq_len(nlevels(se.filt$histology)), inset=0.05)
```

# Differential expression

We can do an analysis to observe differential expression based on the
two following groupings:

***Histology***

We can separate the samples' Max. Histology in two groups: Dysplasia vs.
No Dysplasia.

The *No Dysplasia* group will contain all samples that are Normal, have
Hyperplasia or Metaplasia.

The *Dysplasia* group will contain all samples with Dysplasia: Mild
Dysplasia, Moderate dysplasia and Severe dysplasia

***COPD***

(Chronic obstructive pulmonary disease)

Samples with COPD vs. samples without COPD

## Analysing Fold-Change

In first instance, a simple way to find Differentially Expressed genes
is to simply compare expression levels between the two groups of
samples.

We can do this using **fold-change**, which compares the mean read
counts using the logarithmic scale to try and stabilize variability
across expression levels.

### COPD

Our two groups of samples are: samples with COPD and samples with No
COPD. We calculate the means of the values of the expression levels.

```{r}
copd <- rowMeans(logCPM[, se.filt$copd_status=="COPD"])
no_copd <- rowMeans(logCPM[, se.filt$copd_status=="No COPD"])
```

Now we can build two plots to compare fold-change between the two groups
(Figure \@ref(fig:foldCOPD)).

```{r foldCOPD, fig.cap="Plot analysing the difference in expression between samples with and without COPD."}
par(mfrow=c(1, 2))
plot(copd, no_copd, xlab="COPD", ylab="No COPD", pch=".", cex=4, las=1)
plot((copd+no_copd)/2, copd-no_copd, xlab="(COPD + No COPD)/2", ylab="COPD - No COPD", pch=".", cex=4, las=1)
```

### Histology

Our two groups of samples are No Dysplasia and Dysplasia. We create a
new column in our SE to dichotomize the histology. We assign "Normal",
"Hyperplasia" and "Metaplasia" to the **No Dysplasia** group, and we
assign "Mild dysplasia", "Moderate dysplasia", and "Severe dysplasia" to
the **Dysplasia** group

```{r}
se.filt$dysplasia[se.filt$histology == "Normal"] <- "No Dysplasia"
se.filt$dysplasia[se.filt$histology == "Hyperplasia"] <- "No Dysplasia"
se.filt$dysplasia[se.filt$histology == "Metaplasia"] <- "No Dysplasia"
se.filt$dysplasia[se.filt$histology == "Mild dysplasia"] <- "Dysplasia"
se.filt$dysplasia[se.filt$histology == "Moderate dysplasia"] <- "Dysplasia"
se.filt$dysplasia[se.filt$histology == "Severe dysplasia"] <- "Dysplasia"
```

```{r}
no_dysplasia <- rowMeans(logCPM[, se.filt$dysplasia=="No Dysplasia"])
dysplasia <- rowMeans(logCPM[, se.filt$dysplasia=="Dysplasia"])
```

Now we can build two plots to compare fold-change between the two groups
(Figure \@ref(fig:foldDysplasia))

```{r foldDysplasia, fig.cap="Plot analysing the difference in expression between samples with and without Dysplasia"}
par(mfrow=c(1, 2))
plot(no_dysplasia, dysplasia, xlab="No Dysplasia", ylab="Dysplasia", pch=".", cex=4, las=1)
plot((no_dysplasia+dysplasia)/2, no_dysplasia-dysplasia, xlab="(No Dysplasia + Dysplasia)/2", ylab="No Dysplasia - Dysplasia", pch=".", cex=4, las=1)
```

We can see the fold-change plot for the Dysplasia comparison (Figure
\@ref(fig:foldDysplasia)) has a wider distribution than the COPD
fold-change plot (Figure \@ref(fig:foldCOPD)), which suggests we'll have
a higher number of Differentially Expressed genes in the Dysplasia DE
Analysis.

## Performing Differential Expression (DE)

We perform a simple assessment of the extent of expression changes and
their associated p-values using the F-test implemented in the
R/Bioconductor package [sva](http://bioconductor.org/packages/sva). We
compare two groups of samples in the following ways.

```{r}
library(sva)
```

### COPD

We compare samples that have COPD with samples that don't have COPD. We
have previously subset the data the way we needed.

We build now the corresponding full and null model matrices.

```{r}
mod <- model.matrix(~ se.filt$copd_status, colData(se.filt))
mod0 <- model.matrix(~ 1, colData(se.filt))
```

Finally, we conduct the F-test implemented in the package `sva` and
examine the amount of differential expression between samples with COPD
and without.

```{r, message=FALSE}
pv <- f.pvalue(assays(se.filt)$logCPM, mod, mod0)
sum(p.adjust(pv, method="fdr") < 0.05)
sum(p.adjust(pv, method="fdr") < 0.1)
```

We obtain `r sum(p.adjust(pv, method="fdr") < 0.05)` differentially
expressed (DE) genes at FDR \< 5%, and
`r sum(p.adjust(pv, method="fdr") < 0.1)` differentially expressed (DE)
genes at FDR \< 10%, In Figure \@ref(fig:pdistCOPD) below we can see the
distribution of the resulting p-values.

```{r pdistCOPD, echo=FALSE, out.width="600px", fig.cap="Distribution of raw p-values for an F-test on every gene between samples with COPD and samples without COPD."}
hist(pv, main="", las=1, xlab="P-Value")
```

We build a table with the subset of
`r sum(p.adjust(pv, method="fdr") < 0.1)` DE genes with FDR \< 10% and
show the top-10 genes with lowest p-value in Table
\@ref(tab:DEgenesCOPD) below.

```{r, message=FALSE}
mask <- p.adjust(pv, method="fdr") < 0.1
DEgenesEGs <- names(pv)[mask]
DEgenesSyms <- mcols(se.filt)[DEgenesEGs, "symbol"]
DEgenesPvalue <- pv[mask]
DEgenesDesc <- mcols(se.filt)[DEgenesEGs, "description"]
DEgenesDesc <- sub(" \\[.+\\]", "", DEgenesDesc)
DEgenesTab <- data.frame(EntrezID=DEgenesEGs,
                         Symbol=DEgenesSyms,
                         Description=DEgenesDesc,
                         "P value"=DEgenesPvalue,
                         stringsAsFactors=FALSE, check.names=FALSE)
DEgenesTab <- DEgenesTab[order(DEgenesTab[["P value"]]), ] ## order by p-value
rownames(DEgenesTab) <- 1:nrow(DEgenesTab)
```

```{=html}
<!--
The following code chunk is hidden because its lines are not that relevant
to the analysis of the data. Their purpose is to dump the table of DE genes
into an HTML table and a CSV file to facilitate the examination of long lists
of DE genes while avoiding to produce a lengthy vignette by showing only the
top-10 DE genes.
--->
```
```{r, echo=FALSE}
## generate full table in a CSV file and store it in the 'doc' directory
## twice, once in 'doc' to enable quickly look up during vignette editing
## and building with 'devtools::build_vignettes()' and a second time in
## 'inst/doc' to make these files available at install.
fnameCSV <- "DEgenesCOPD.csv"
fpathCSV <- file.path(path2pkg, "doc", fnameCSV)
write.csv(DEgenesTab, fpathCSV, row.names=FALSE)
fpathCSV <- file.path(path2pkg, "inst", "doc", fnameCSV)
write.csv(DEgenesTab, fpathCSV, row.names=FALSE)

## generate full table in HTML and store it into the 'doc' directory
## twice, just as we did with the CSV file. note that because the
## table caption is not translated from Markdown, but directly copied
## into HTML, we need to avoid using the '<' symbol, as in FDR < 10%,
## and put its HTML code instead (&lt;)
ktab <- kable(DEgenesTab, "html", escape=FALSE, row.names=TRUE,
              caption=sprintf("Differentially expressed genes. Differentially expressed genes between samples with COPD and samples without COPD, with FDR &lt; 10%% (CSV <a href=\"%s\" download>file</a>).",
                              fnameCSV))
ktab <- kable_styling(ktab,
                      bootstrap_options=c("stripped", "hover", "responsive"),
                      fixed_thead=TRUE)
fnameHTML <- "DEgenesCOPD.html"
fpathHTML <- file.path(path2pkg, "doc", fnameHTML)
save_kable(ktab, file=fpathHTML, self_contained=TRUE)
fpathHTML <- file.path(path2pkg, "inst", "doc", fnameHTML)
save_kable(ktab, file=fpathHTML, self_contained=TRUE)
```

```{=html}
<!--
The following code chunk is also hidden because its purpose is to produce
a short HTML table of the top-10 DE genes. The reason show here only the
top-10 genes is to producing a long vignette. The full HTML table and the
corresponding CSV file are linked through the caption of this short HTML
table.

note that because the table caption is not translated from Markdown, but
directly copied into HTML, we need to avoid using the '<' symbol, as in
FDR < 10%, and put its HTML code instead (&lt;)
--->
```
```{r DEgenesCOPD, echo=FALSE}
ktab <- kable(DEgenesTab[1:10, ], "html", escape=FALSE, row.names=TRUE, 
              caption=sprintf("Differentially expressed genes. Top-10 differentially expressed genes with lowest p-value between samples with COPD and without COPD with FDR &lt; 10%%. To see the full list of DE genes, follow this <a href=\"%s\" target=\"_blank\">link</a> or download this CSV <a href=\"%s\" download>file</a>.",
                              fnameHTML, fnameCSV))
kable_styling(ktab, position="center")
```

### Histology

We compare samples that have Dysplasia with samples that don't have
Dysplasia. We have previously subset the data the way we needed.

We build now the corresponding full and null model matrices.

```{r}
mod <- model.matrix(~ se.filt$dysplasia, colData(se.filt))
mod0 <- model.matrix(~ 1, colData(se.filt))
```

Finally, we conduct the F-test implemented in the package `sva` and
examine the amount of differential expression between samples with
Dysplasia and samples with Hyperplasia/Normal samples.

```{r, message=FALSE}
pv <- f.pvalue(assays(se.filt)$logCPM, mod, mod0)
sum(p.adjust(pv, method="fdr") < 0.05)
```

We obtain `r sum(p.adjust(pv, method="fdr") < 0.05)` differentially
expressed (DE) genes at FDR \< 5% and
`r sum(p.adjust(pv, method="fdr") < 0.1)` at FDR \< 10%. In Figure
\@ref(fig:pdistDysplasia) below we can see the distribution of the
resulting p-values.

```{r pdistDysplasia, echo=FALSE, out.width="600px", fig.cap="Distribution of raw p-values for an F-test on every gene between samples with Dysplasia and samples with Hyperplasia/Normal samples."}
sum(p.adjust(pv, method="fdr") < 0.05)
hist(pv, main="", las=1, xlab="P-Value")
```

We build a table with the subset of
`r sum(p.adjust(pv, method="fdr") < 0.05)` DE genes with FDR \< 5% and
show the top-10 genes with lowest p-value in Table
\@ref(tab:DEgenesDysplasia) below.

```{r, message=FALSE}
mask <- p.adjust(pv, method="fdr") < 0.05
DEgenesEGs <- names(pv)[mask]
DEgenesSyms <- mcols(se.filt)[DEgenesEGs, "symbol"]
DEgenesPvalue <- pv[mask]
DEgenesDesc <- mcols(se.filt)[DEgenesEGs, "description"]
DEgenesDesc <- sub(" \\[.+\\]", "", DEgenesDesc)
DEgenesTab <- data.frame(EntrezID=DEgenesEGs,
                         Symbol=DEgenesSyms,
                         Description=DEgenesDesc,
                         "P value"=DEgenesPvalue,
                         stringsAsFactors=FALSE, check.names=FALSE)

DEgenesTab <- DEgenesTab[order(DEgenesTab[["P value"]]), ] ## order by p-value
rownames(DEgenesTab) <- 1:nrow(DEgenesTab)
```

```{=html}
<!--
The following code chunk is hidden because its lines are not that relevant
to the analysis of the data. Their purpose is to dump the table of DE genes
into an HTML table and a CSV file to facilitate the examination of long lists
of DE genes while avoiding to produce a lengthy vignette by showing only the
top-10 DE genes.
--->
```
```{r, echo=FALSE}
## generate full table in a CSV file and store it in the 'doc' directory
## twice, once in 'doc' to enable quickly look up during vignette editing
## and building with 'devtools::build_vignettes()' and a second time in
## 'inst/doc' to make these files available at install.
fnameCSV <- "DEgenesDysplasia.csv"
fpathCSV <- file.path(path2pkg, "doc", fnameCSV)
write.csv(DEgenesTab, fpathCSV, row.names=FALSE)
fpathCSV <- file.path(path2pkg, "inst", "doc", fnameCSV)
write.csv(DEgenesTab, fpathCSV, row.names=FALSE)

## generate full table in HTML and store it into the 'doc' directory
## twice, just as we did with the CSV file. note that because the
## table caption is not translated from Markdown, but directly copied
## into HTML, we need to avoid using the '<' symbol, as in FDR < 10%,
## and put its HTML code instead (&lt;)
ktab <- kable(DEgenesTab, "html", escape=FALSE, row.names=TRUE,
              caption=sprintf("Differentially expressed genes. Differentially expressed genes between samples with Dysplasia and samples with no Dysplasia, with FDR &lt; 5%% (CSV <a href=\"%s\" download>file</a>).",
                              fnameCSV))
ktab <- kable_styling(ktab,
                      bootstrap_options=c("stripped", "hover", "responsive"),
                      fixed_thead=TRUE)
fnameHTML <- "DEgenesDysplasia.html"
fpathHTML <- file.path(path2pkg, "doc", fnameHTML)
save_kable(ktab, file=fpathHTML, self_contained=TRUE)
fpathHTML <- file.path(path2pkg, "inst", "doc", fnameHTML)
save_kable(ktab, file=fpathHTML, self_contained=TRUE)
```

```{=html}
<!--
The following code chunk is also hidden because its purpose is to produce
a short HTML table of the top-10 DE genes. The reason show here only the
top-10 genes is to producing a long vignette. The full HTML table and the
corresponding CSV file are linked through the caption of this short HTML
table.

note that because the table caption is not translated from Markdown, but
directly copied into HTML, we need to avoid using the '<' symbol, as in
FDR < 10%, and put its HTML code instead (&lt;)
--->
```
```{r DEgenesDysplasia, echo=FALSE}
ktab <- kable(DEgenesTab[1:10, ], "html", escape=FALSE, row.names=TRUE, 
              caption=sprintf("Differentially expressed genes. Top-10 differentially expressed genes with lowest p-value between samples with Dysplasia and samples with No Dysplasia, with FDR &lt; 5%%. To see the full list of DE genes, follow this <a href=\"%s\" target=\"_blank\">link</a> or download this CSV <a href=\"%s\" download>file</a>.",
                              fnameHTML, fnameCSV))
kable_styling(ktab, position="center")
```

We can also see, when comparing both *Distribution of raw p-values for
an F-test* plots (Figures \@ref(fig:pdistCOPD) and
\@ref(fig:pdistDysplasia)), that the one corresponding to the Dysplasia
grouping has a lot more of p-values closer to 0, which suggests that
grouping will have more DE genes.

# Functional analysis

Here we will do the functional analysis.

# Discussion

Here we discuss the findings.

# Conclusions

Here we summarize our conclusions.

# Session information

```{r}
sessionInfo()
```

# References
