---
title: "IEO Project Analysis"
author:
- name: "Radostina Kisleva, Emma Juan Salazar, Tamara-Dora Veres"
  affiliation: Universitat Pompeu Fabra
  email: radostina.kisleva01@estudiant.upf.edu,  emma.juan01@estudiant.upf.edu,  tamaradora.veres01@estudiant.upf.edu
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
    fig_captions: yes
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{IEOprojectAnalysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!--
In this first chunk of code, which will not be shown in the resulting
document (echo=FALSE) sets up global processing options, such as whether
a comment character should appear before code results (set to the null
character string in this case), collapse source and output blocks into
a single code block (collapse=TRUE), align figures to the center
(fig.align="center") or cache the results to speed up vignette building
(cache=FALSE thus disabled in this case). A full description of possible
options can be found at http://yihui.name/knitr/options
--->

```{r setup, echo=FALSE, cache=FALSE}
library(knitr) ## kable()
library(kableExtra) ## kable_styling(), save_kable()
library(here) ## here()
library(usethis) ## use_directory()

knitr::opts_chunk$set(
  collapse=TRUE,
  comment="",
  fig.align="center",
  fig.wide=TRUE,
  cache=FALSE
)

## this option avoid use_directory() being verbose
options(usethis.quiet=TRUE)

## create these paths at build time if they do not exist
use_directory(file.path("doc"))
use_directory(file.path("inst", "doc"))

## fetch the package root directory
path2pkg <- here()

```

# Introduction

Lung cancer is the leading cause of cancer-related death in the United
States. The molecular events preceding the onset of disease are poorly
understood, and no effective tools exist to identify smokers with
premalignant lesions (PMLs) that will progress to invasive cancer.
Previous research has identified molecular changes in the smoke-exposed
airways. However, the focus of the paper "Detecting the Presence and
Progression of Premalignant Lung Lesions via Airway Gene
Expression" @Beane17 shifts the focus to the first stages of the disease progress.
They utilize samples from the same
airway field of injury to study PMLs and their potential in preventing
lung cancer.

The researchers' goal is to profile samples extracted from normal
appearing bronchial mucosa by mRNA-Seq, and attempt to find
differentially expressed gene sets and pathways between subjects with
PMLs and without PMLs. From the original 82 samples, 50 subjects had
PML, 25 didn't, and 7 were excluded in the end.

The resulting raw RNA-seq data has been deposited at the Gene
Expression Omnibus (GEO), where they are publicly available under
accession
[GSE79209](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE79209).
Here, we show a first exploratory analysis of the corresponding RNA-seq
gene expression profiles generated as a table of counts using the DEE2
(<https://dee2.io>) pipeline by @ziemann19, and further packaged into a
[SummarizedExperiment](https://bioconductor.org/packages/SummarizedExperiment)
object with genes mapped to Entrez identifiers. This object also stores
the phenotypic information about the profiled samples that has been also
made available at GEO.

# Quality assessment

## Data import and cleaning

We start importing the raw table of counts.

<!--
The option 'message=FALSE' avoid dumping R messages such as "Loading required package: methods"
into the output of the report.
-->

```{r, message=FALSE}
library(SummarizedExperiment)

se <- readRDS(file.path(system.file("extdata",
                                    package="IEOproject"),
                        "GSE79209.rds"))
se

```

We have `r nrow(se)` genes by `r ncol(se)` samples. From the first row
and column names shown by the object, we can figure out that genes are
defined by [Entrez](https://www.ncbi.nlm.nih.gov/gene) [@maglott10]
identifiers and samples by Sequence Read Archive Run
([SRR](https://www.ncbi.nlm.nih.gov/books/NBK56913/#search.what_do_the_different_sra_accessi)) identifiers.

The row data in this object contains information about the profiled genes.

```{r}
head(rowData(se))
```
Among this information, the gene symbol and description are potentially
useful for interpreting results of, for instance, a differential
expression analysis.

### Exploratory Data Analysis

Let's explore now the information we have for the samples: phenotypic
and technical data.

Doing EDA will allow us to identify how many phenotypic variables we
have, the technical variables, and see what could interest us later for
Experimental Design.

```{r}
dim(colData(se))
head(colData(se), n=3)
```

We have a total of `r ncol(colData(se))` phenotypic variables.

A first approach can be to simply analyze all of the column names to get
a first sense of what we have:

```{r}
colnames(colData(se))
```

The second column `geo_accession` contains GEO Sample Accession Number
([GSM](https://www.ncbi.nlm.nih.gov/geo/info/overview.html)) identifers.
GSM identifiers define individual samples, understood in our context as
individual sources of RNA. We can check if we have any technical replicates
as follows:

```{r}
length(unique(se$geo_accession))
table(lengths(split(colnames(se), se$geo_accession)))
```

So, we have `r length(unique(se$geo_accession))` different individual
samples which matches the number of samples that we have `r ncol(se)`.
Therefore we don't have any technical replicates and we can proceed with
the next step.

To proceed further exploring this dataset, we are going to use the
[edgeR](https://bioconductor.org/packages/edgeR) package and build
a `DGEList` object, incorporating the gene metadata, which includes
the gene symbol.

```{r, message=FALSE}
library(edgeR)

dge <- DGEList(counts=assays(se)$counts, genes=rowData(se))
head(dge$samples, 3)
dim(dge)
```
Calculate $\log_2$ CPM units of expression and put them as an
additional assay element to ease their manipulation. This provides a
standarized and normalized representation of gene expression data.

```{r}
assays(se)$logCPM <- cpm(dge, log=TRUE)
assays(se)$logCPM[1:5, 1:5]
```

Let's explore now some of the phenotypic variables.

To identify the columns that will be useful to us, we have to see the
length of the levels of each variable (how many unique values there are)
to see what can be used to make groups of samples. If a variable has 81
levels, or only 1, we are not interested in it. Therefore we want to
obtain a list of all the columns that have a different number than
those.

```{r}
interesting_se <- data.frame(row.names=se$title)
technical_se <- data.frame(row.names=se$title)
for(column in colnames(colData(se))){
  se[[column]] <- factor(se[[column]])
  if(length(levels(se[[column]]))>1 && length(levels(se[[column]]))<81){
    interesting_se[[column]] <- se[[column]]
  } else{
    technical_se[[column]] <- se[[column]]
  }
}
```

We can now look at the levels of each of these variables to see what
type of data they offer

```{r}
head(interesting_se)
```

We can see we have the following 8 phenotypic (or environmental)
variables (some information is repeated):

**Age** (characteristics_ch1.1)

**Sex** (characteristics_ch1.2)

**Smoking status** (characteristics_ch1.3): Current smoker or Former
smoker

**Pack years** (characteristics_ch1.4): how many cigarettes a person has
smoked in their lifetime

**COPD Status** (characteristics_ch1.1): Chronic obstructive pulmonary
disease

**Mean GC Content** (characteristics_ch1.18)

**Max Histology** (characteristics_ch1.6): Histology types (tissue
characteristics)

-   This variable determines what they'll call dysplasia status (see
    next variable). Histology is used to determine wether a sample has
    Premalignant Lesions or not, the two main groups of the paper.
    -   The PML group (**dysplasia**) includes samples that had:

        -   Mild Dysplasia

        -   Moderate Dysplasia

        -   Severe Dysplasia

    -   The No PML group (**no dysplasia**) includes samples that had:

        -   Hyperplasia

        -   Normal histology

    -   Samples that had **metaplasia** were excluded from the analysis

**Dysplasia Status** (characteristics_ch1.7)

-   Dysplasia -\> PML

-   No Dysplasia -\> No PML

-   Metaplasia -\> excluded

To facilitate handling this variables we are going to perform the following
modifications:

```{r}
se$sex <- gsub("Sex: ", "", se$characteristics_ch1.2)
se$sex <- factor(se$sex, levels = c("Male", "Female"))

se$smoker <- gsub("smoking status: ", "", se$characteristics_ch1.3)
se$smoker <- factor(se$smoker, levels = c("Current smoker", "Former smoker"))

se$copd_status <- gsub("copd status: ", "", se$characteristics_ch1.5)
se$copd_status <- factor(se$copd_status, levels = c("COPD", "No COPD"))

se$histology <- gsub("max histology: ", "", se$characteristics_ch1.6)
se$histology <- factor(se$histology, levels(se$histology) <- c("Hyperplasia", "Metaplasia", "Mild dysplasia", "Moderate dysplasia", "Normal", "Severe dysplasia"))

se$dysplasia <- gsub("Subject L.*, ", "", se$title)
se$dysplasia <- factor(se$dysplasia)
```

To faciliate working with the pack years variable will split it into four 
categorical groups - Low, Medium, High and Very high. 

To do that we will:  
1. Calculate quartiles for the data.  
2. Assign categories based on quartile ranges.  
3. Convert the values into their corresponding categories.  


```{r}
se$characteristics_ch1.4 <- gsub("pack years: ", "", se$characteristics_ch1.4)
class(se$characteristics_ch1.4)
se$characteristics_ch1.4 <- as.numeric(se$characteristics_ch1.4)

quartiles <- quantile(se$characteristics_ch1.4, probs = c(0.25, 0.5, 0.75))
```

Now that we have the quartiles, we can proceed to assign categories to the values based on these quartiles.

Values below the first quartile (Q1) will be categorized as "low."
Values between Q1 and the median (Q2) will be categorized as "medium."
Values between the median (Q2) and the third quartile (Q3) will be categorized as "high."
Values above the third quartile (Q3) will be categorized as "very high."


```{r}

se$pack_years <- cut(se$characteristics_ch1.4,
                          breaks = c(-Inf, quartiles[1], quartiles[2], quartiles[3], Inf),
                          labels = c("Low", "Medium", "High", "Very High"),
                          include.lowest = TRUE)

```

To faciiate working with the age variable, we'll split it into four
intervals

To do that we will:\
1. Calculate quartiles for the data.\
2. Assign categories based on quartile ranges.\
3. Convert the values into their corresponding categories.

```{r}

se$age <- gsub("age: ", "", se$characteristics_ch1.1)
class(se$age)
se$age <- as.numeric(se$age)

quartiles <- quantile(se$age, probs = c(0.25, 0.5, 0.75))
quartiles

```

Now that we have the quartiles, we can proceed to assign categories to
the values based on these quartiles.

Values below the first quartile (Q1) will be categorized as "low."
Values between Q1 and the median (Q2) will be categorized as "medium."
Values between the median (Q2) and the third quartile (Q3) will be
categorized as "high." Values above the third quartile (Q3) will be
categorized as "very high."

```{r}
se$age_intervals <- cut(se$age,
                          breaks = c(-Inf, quartiles[1], quartiles[2], quartiles[3], Inf),
                          labels = c("to 58", "to 64", "to 69", "+ 69"),
                          include.lowest = TRUE)

```
In Table \@ref(tab:pheno) below, we show the five variables of interest
(smoker status, pack years, histology, dysplasia and COPD status)
jointly to gather as much understanding as possible on the underlying
experimental design.

```{r pheno, echo=FALSE, message=FALSE}
tmpdf <- data.frame(Identifer=colnames(se),
                    Age=se$age_intervals,
                    Sex=se$sex,
                    "Smoker status"=se$smoker,
                    "Pack Years"=se$pack_years,
                    "COPD Status"=se$copd_status,
                    Histology=se$histology,
                    Dysplasia=se$dysplasia,
                    check.names=FALSE)
ktab <- kable(tmpdf, caption="Phenotypic variables.")
kable_styling(ktab, position="center")
```

## Sequencing depth

In the following step we will evaluate the sequencing depth in terms of the total number of read counts that are mapped to each sample's genome. The sequencing depth per sample, commonly referred to as library sizes, is displayed in Figure \@ref(fig:libsizes) below in increasing order. 
<!---
you can control the height and width in pixels of the figure with
'out.height' and 'out.width'. Figures are automatically numbered,
to refer to them in the main text you should use the notation shown
above as \@ref(fig:xxxx) with xxxx being the label in the code chunk
that also gives the filename of the figure. This name must be unique
--->

```{r libsizes, echo=FALSE, height=8, width=8, out.width="600px", fig.cap="Library sizes in increasing order."}
par(mar=c(7, 5, 2, 2))
ord <- order(dge$sample$lib.size/1e6)
ordmreads <- dge$sample$lib.size[ord]/1e6
names(ordmreads) <- colnames(se)[ord]
bp <- barplot(ordmreads, las=1, ylab="Millions of reads",
              xlab="", col=c("blue", "green", "red", "purple", "orange", "pink")[factor(se$histology[ord])], las=2, ylim = c(0, 50))
legend("topleft", c("Hyperplasia", "Metaplasia", "Mild dysplasia", "Moderate dysplasia", "Normal", "Severe dysplasia"), fill=c("blue", "green", "red", "purple", "orange", "pink"), inset=0.01, cex=0.85)

```

The plot displays a range of sequencing depths for the samples, each represented by an SRR identifier. The sequencing depth across our samples ranges from 20 to 50 million reads, with the majority falling between 30 to 40 million reads. This indicates a generally high level of coverage, which is beneficial for accurate genomic analysis. The plot also suggests that the highest number of reads corresponds to mild and moderate dysplasia, indicating early stages of cellular abnormalities that could potentially progress to cancer. The predominance of mild dysplasia in these samples could be of particular interest for early detection and intervention studies. There does not appear to be any evident bias in terms of sequencing depth across the conditions represented.



## Distribution of expression levels among samples

Figure \@ref(fig:distRawExp) below shows the distribution of expression
values per sample in logarithmic CPM units of expression.

<!---
the option echo=FALSE hides the R code. When plotting in general one
does not want to see the code. Options fig.height and fig.width control
height and width of the plot in inches while out.height and out.width
do it in the final output file; see http://yihui.name/knitr/options for
full details.
--->

```{r distRawExp, echo=FALSE, fig.height=5, fig.width=5, out.width="600px", fig.cap="Non-parametric density distribution of expression profiles per sample.", message=FALSE}
library(geneplotter)
par(mar=c(4, 5, 1, 1))
lst <- as.list(as.data.frame(assays(se)$logCPM))
multidensity(lst, xlab="log 2 CPM", legend=NULL,
             main="", las=1)
```
There are no substantial differences between the samples in the
distribution of expression values.

## Distribution of expression levels among genes

Let's calculate now the average expression per gene through all the
samples. Figure \@ref(fig:exprdist) shows the distribution of those
values across genes.

```{r exprdist, echo=FALSE, out.width="600px", fig.cap="Distribution of average expression level per gene."}
avgexp <- rowMeans(assays(se)$logCPM)
hist(avgexp, xlab="log2 CPM", main="", las=1)
```
The histogram displays a bimodal distribution, which means there are two distinct peaks. The large peak to the left is the most prominent feature of the histogram, showing that most genes in this dataset are expressed at low levels or possibly not at all, under the conditions of the experiment. On the right part, the peak which indicates that as gene expression levels increase, fewer genes are found at these higher expression levels.


## Filtering of lowly-expressed genes

To filter out lowly-expressed genes, we first calculate the row means of the expression data and apply a logical mask to remove genes with an average logCPM ≤ 1. This step refines the dataset and removes noise, ensuring subsequent analysis is based on high-quality data.

Next, we calculate a CPM cutoff based on the smallest library size. We check for any missing values in the histology data to ensure data integrity.

We then focus on filtering genes based on their presence in samples with premalignant lesions (PML), specifically those categorized as "Mild dysplasia", "Moderate dysplasia", and "Severe dysplasia". We count the number of PML samples and calculate the CPM for these samples only.

Finally, we apply a logical mask to retain genes that meet the CPM cutoff in at least the number of PML samples. This two-step filtering process is crucial for refining the dataset and focusing the analysis on genes with relevant expression patterns. Figure \@ref(fig:lowexpgenes) shows the distribution of expression values across genes before and after filtering.

```{r}
mask_initial <- rowMeans(assays(se)$logCPM) > 1
se_initial_filt <- se[mask_initial, ]
dge_initial_filt <- dge[mask_initial, ]
dim(se_initial_filt)

cpmcutoff <- round(10/min(dge$samples$lib.size/1e6), digits=1)
cpmcutoff

if(any(is.na(se$histology))) {
  stop("Missing values found in the histology column.")
}

pml_status <- se$histology %in% c("Mild dysplasia", "Moderate dysplasia", "Severe dysplasia")
nsamplescutoff <- sum(pml_status)
nsamplescutoff

pml_indices <- which(pml_status)

cpm_pml <- cpm(dge_initial_filt)[, pml_indices]

mask_final <- rowSums(cpm_pml > cpmcutoff) >= nsamplescutoff
sum(mask_final)
se.filt <- se_initial_filt[mask_final, ]
dge.filt <- dge_initial_filt[mask_final, ]
dim(se.filt)
```

```{r lowexpgenes, echo=FALSE, out.width="600px", fig.cap="Distribution of lowly-expressed genes."}
par(mar=c(4, 5, 1, 1))
avgexp <- rowMeans(assays(se)$logCPM)
h <- hist(avgexp, xlab=expression("Expression level (" * log[2] * "CPM)"),
          main="", las=1, col="grey", cex.axis=1.2, cex.lab=1.5)
x <- cut(rowMeans(assays(se.filt)$logCPM), breaks=h$breaks)
lines(h$mids, table(x), type="h", lwd=10, lend=1, col="red")
legend("topright", c("All genes", "Filtered genes"), fill=c("grey", "red"))

```

We are left with `r nrow(se.filt)` genes.

## Normalization

We calculate now the normalization factors on the filtered expression
data set.

```{r}
dge.filt <- calcNormFactors(dge.filt)
```

Replace the raw log2 CPM units in the corresponding assay element of
the `SummarizedExperiment` object, by the normalized ones.

```{r}
assays(se.filt)$logCPM <- cpm(dge.filt, log=TRUE,
                              normalized.lib.sizes=TRUE)
```

## MA-plots

We examine now the MA-plots of the normalized expression profiles
in the figures below.

<!---
Here we make a MA-plot for each sample. The options 'fig.height'
and 'fig.width' control the relative image size in *inches*. The
final image size results from 'height'x'dpi' and 'width'x'dpi',
where 'dpi' is the image resolution in "dots per inch" (by default
dpi=72). To scale the image to a desired size use 'out.width' and
'out.height'. More information at http://yihui.name/knitr/options
--->
```{r maPlots, fig.height=18, fig.width=10, dpi=100, echo=FALSE, fig.cap="MA-plots of filtered and normalized expression values."}
par(mfrow=c(5, 4), mar=c(4, 5, 3, 1))
for (i in 1:ncol(se.filt)) {
  A <- rowMeans(assays(se.filt)$logCPM)
  M <- assays(se.filt)$logCPM[, i] - A
  smoothScatter(A, M, main=colnames(se.filt)[i], las=1)
  abline(h=0, col="blue", lwd=2)
  lo <- lowess(M ~ A)
  lines(lo$x, lo$y, col="red", lwd=2)
}
```

For genes above the horizontal line at M=0, expression is higher in the condition being tested against the reference, as for genes below, expression is lower.
The red line represents a loess fit to the data, which should ideally be around M=0 if there is no differential expression between conditions. Any systematic deviation from M=0 indicates potential bias or systematic variation in the data. In most of the plots, the red line remains close to M=0 across the range of A values, indicating no systematic bias in expression changes relative to gene abundance.
Some plots show a slight dip in the loess line at the lower end of the A scale, suggesting a potential bias for lowly expressed genes which could be potential candidates for differential expression. These points would be of interest for follow-up analyses.

## Experimental design and batch identification

To identify potential columns which could constitute unwanted variability, we looked at columns that contain information related to technical aspects, experimental conditions or processing details that could vary between different batches. 

When doing EDA, we created a dataframe with all the phenotypical
variables, and another with all the technical ones.

```{r}
head(technical_se)
```

Here are some columns which we thought could introduce unwanted variability: `submission_date`, `last_update_date`, `type`, `channel_count`, `extract_protocol_ch1`, `extract_protocol_ch1.1`, `data_processing`, `data_processing.1`, `data_processing.2`, `data_processing.3`, `data_processing.4`, `data_processing.5`, `data_processing.6`, `platform_id`, `instrument_model`, `library_selection`, `library_source`, `library_strategy`.

These variables have the same values across all samples which suggests that they are not contributing to the variability of the dataset. For example:
```{r}
table(se.filt$histology, se.filt$library_strategy)
```

There is perfect correlation between library strategy and histology. This means that differences between our samples will be due to biological differences rather than technical. 

We can discard all of the variables that have the same value across all
samples:

```{r}
for(column in colnames(technical_se)){
  if(length(levels(technical_se[[column]]))==1){
    technical_se[[column]] <- NULL
  }
}
head(technical_se)
```

The rest of the variables are related to technical aspects
related to the sequencing that's specific to every sample, therefore we can discard them as sources of unwanted variability.

### Phenotypical Variables

We have to now take a look at how our phenotypical variables interact
and wether they might explain any sample clustering we may find, to help
us make a decision on what groups to choose for our DE analysis and
their covariates,

First, let's examine the distribution between **dysplasia** (our outcome
of interest, seeing as the paper's goal is to find DE genes between
samples with PML and without) and [histology]{.underline}.

```{r}
table(se.filt$dysplasia, se.filt$histology)
```

This is a very obvious result, since we know dysplasia was drawin from
the histology variable. This is a great example of combinations that
aren't sequenced (in this case, because they can't exist.)

It may be more interesting to see what happens between our outcome of
interest and variables that are independent from it.

We can also examine the distribution between histology and our other
phenotypic variables

```{r}
table(se.filt$histology, se.filt$age_intervals)
table(se.filt$histology, se.filt$sex)
table(se.filt$histology, se.filt$pack_years)
table(se.filt$histology, se.filt$copd_status)
```

We can see that not all combinations of pack-years and histology have
been sequenced. Some combinations have missing values, indicated by 0s.

It's interesting to note that the 0s indicate some interesting remarks:

-   Groups like low or medium pack years (people that have smoked less
    cigarettes over the course of their lives) have no samples with
    severe dysplasia, meaning severe detriment to their bronchial
    tissue.

-   There's also no females with severe dysplasia.

-   And the severe dysplasia group also lacks samples on the younger age
    interval.

The distribution we're interested in, however, is the one between these
phenotypic variables and the Dysplasia variable (same concept as
histology, but grouped in three groups)

```{r}
table(se.filt$dysplasia, se.filt$age_intervals)
table(se.filt$dysplasia, se.filt$sex)
table(se.filt$dysplasia, se.filt$pack_years)
table(se.filt$dysplasia, se.filt$copd_status)
```

We can see that we have sequences for all of these different
combinations.

### Hierarchical Clustering

We examine now how samples group together by hierarchical clustering and
multidimensional scaling, annotating pack-years and histology. We
calculate again log CPM values with a high prior count (3) to moderate
extreme fold-changes produced by low counts. The resulting dendrogram is
shown in Figure \@ref(fig:sampleClustering).

```{r sampleClustering, fig.height=8, fig.width=10, dpi=100, echo=FALSE, fig.cap="Figure S6: Hierarchical clustering of the samples. Labels correspond to pack-years and sample identifer, while colors indicate histology groups."}
logCPM <- cpm(dge.filt, log=TRUE, prior.count=3)
d <- as.dist(1-cor(logCPM, method="spearman"))
sampleClustering <- hclust(d)
batch <- as.integer(se.filt$histology)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- paste(se.filt$dysplasia)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
function(x, batch, labels) {
  if (is.leaf(x)) {
    attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x,"label")]))
    attr(x, "label") <- as.vector(labels[attr(x, "label")])
    }
  x
  }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples", cex=1)
legend("topright", levels(se.filt$histology),
       fill=seq_len(nlevels(se.filt$histology)))

```

Samples are linked together based on their similarity, with the branches
connecting those most similar first. As we move up the tree, the
connections represent progressively less similar groups until all are
joined in a single cluster. Samples do not really cluster together by
histology type, which means histology type is not the primary factor
driving the similarity or dissimilarity between samples.

If we look more closely at the labels, we can also see even if we look
at clustering through the 3 dysplasia groups instead of all 6 histology
groups, there is still no clear clustering.

We can now try to identify a phenotypical variable that will justify
clustering

```{r}
# Age plot
batch <- as.integer(se.filt$age_intervals)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- paste(se.filt$dysplasia)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
function(x, batch, labels) {
  if (is.leaf(x)) {
    attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x,"label")]))
    attr(x, "label") <- as.vector(labels[attr(x, "label")])
    }
  x
  }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples - Age", cex=1)
legend("topright", levels(se.filt$age_intervals),
       fill=seq_len(nlevels(se.filt$age_intervals)))

# Sex plot
batch <- as.integer(se.filt$sex)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- paste(se.filt$dysplasia)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
function(x, batch, labels) {
  if (is.leaf(x)) {
    attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x,"label")]))
    attr(x, "label") <- as.vector(labels[attr(x, "label")])
    }
  x
  }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples - Sex", cex=1)
legend("topright", levels(se.filt$sex),
       fill=seq_len(nlevels(se.filt$sex)))

# COPD plot
batch <- as.integer(se.filt$age_intervals)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- paste(se.filt$dysplasia)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
function(x, batch, labels) {
  if (is.leaf(x)) {
    attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x,"label")]))
    attr(x, "label") <- as.vector(labels[attr(x, "label")])
    }
  x
  }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples - COPD", cex=1)
legend("topright", levels(se.filt$age_intervals),
       fill=seq_len(nlevels(se.filt$age_intervals)))

# Smoking status
batch <- as.integer(se.filt$smoker)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- paste(se.filt$dysplasia)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
function(x, batch, labels) {
  if (is.leaf(x)) {
    attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x,"label")]))
    attr(x, "label") <- as.vector(labels[attr(x, "label")])
    }
  x
  }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples - Smoking status", cex=1)
legend("topright", levels(se.filt$smoker),
       fill=seq_len(nlevels(se.filt$smoker)))

# Pack years
batch <- as.integer(se.filt$pack_years)
sampleDendrogram <- as.dendrogram(sampleClustering, hang=0.1)
names(batch) <- colnames(se.filt)
outcome <- paste(se.filt$dysplasia)
names(outcome) <- colnames(se.filt)
sampleDendrogram <- dendrapply(sampleDendrogram,
function(x, batch, labels) {
  if (is.leaf(x)) {
    attr(x, "nodePar") <- list(lab.col=as.vector(batch[attr(x,"label")]))
    attr(x, "label") <- as.vector(labels[attr(x, "label")])
    }
  x
  }, batch, outcome)
plot(sampleDendrogram, main="Hierarchical clustering of samples - Pack Years", cex=1)
legend("topright", levels(se.filt$pack_years),
       fill=seq_len(nlevels(se.filt$pack_years)))
```

No clustering is evident in any of the resulting plots except for
Smoking Status, where we can see some clustering between the two
categories: Current smoker vs. Former smoker. A different plot may give
us more insight into these clusters.

In Figure \@ref(fig:mdsPlot) we show the corresponding MDS plot for the
Smoking status variable.

```{r mdsPlot, fig.height=5, fig.width=8, dpi=100, echo=FALSE, fig.cap="Figure S7: Multidimensional scaling plot of the samples. Labels correspond to treatment and colors indicate sample group."}
library(ggplot2)
library(RColorBrewer)

outcome <- se.filt$smoker
outcome2 <- se.filt$dysplasia
names(outcome) <- colnames(se.filt)
palette <- brewer.pal(3, "Set2")

plotMDS(dge.filt, labels=outcome2, col=palette[as.numeric(outcome)])
legend("bottomleft",
       levels(se.filt$smoker),
       fill=palette[seq_len(nlevels(outcome))],
       inset=0.05)
```

We can distinguish more clearly that there are indeed two clusters, one
for Current smokers and another for Former smokers. This leads us to
assume these two groups have significant differences in their gene
expression.

# Differential expression

The next step is to do a Differential Expression Analysis. At first
we'll simply use fold-change values to get a first glance at
differentially expressed genes, and we'll follow that by DE through
linear regression. We chose to do two separate DE Analysis, based on the
following variables:

**PML (Premalignant Lesions)**

Samples with dysplasia vs. samples with no dysplasia.

The goal of this DE is to reproduce the findings in the paper

**Smoking status**

From the clustering we've been able to tell there is a clear difference
between current smokers and former smokers. We think it's meaningful
enough to warrant doing another Differentian Expression Analysis

## Fold Change to Analyze Differential Expression

In first instance, a simple way to find Differentially Expressed genes
is to simply compare expression levels between the two groups of
samples.

We can do this using **fold-change**, which compares the mean read
counts (using the logarithmic scale to try and stabilize variability
across expression levels).

We'll see how many genes change their expression between the two groups
using two different classifications.

### PML (Dysplasia)

The two main groups we want to crontrast are samples with premalignant
legions (PML) and samples without them. Thas is why our oucome target is
Dysplasia.

```{r}
no_dysplasia <- rowMeans(logCPM[, se.filt$dysplasia=="no dysplasia"])
dysplasia <- rowMeans(logCPM[, se.filt$dysplasia=="dysplasia"])
```

We can now calculate fold-change values for each gene in these two
groups.

```{r}
log2fc_dysplasia <- dysplasia - no_dysplasia
ranking <- order(abs(log2fc_dysplasia), decreasing=TRUE)
fold_change_dysplasia <- data.frame(EntrezID=names(log2fc_dysplasia[ranking]),
                                    Log2FC=round(log2fc_dysplasia[ranking], digits=3),
                                    FC=round(2^log2fc_dysplasia[ranking], digits=3),
                                    "1/FC"=round(2^(-log2fc_dysplasia[ranking]), digits=3),
                                    row.names=rowData(se.filt)$symbol[ranking],
                                    check.names=FALSE)
de_fc_dysplasia <- head(fold_change_dysplasia, n=10)
de_fc_dysplasia
```

In the resulting Data Frame:

-   If a gene has a **positive** Log2FC value, it means said gene is
    **FC** times more upregulated in samples with *Dysplasia* compared
    to in samples *without Dysplasia*.

-   If a gene has a **negative** Log2FC value, it means said gene is
    **1/FC** times more upregulated in patients *without Dysplasia*
    compared to in patients with *Dysplasia*

    -   For example, CLDN18 is more upregulated in samples with
        Dysplasia than in samples without Dysplasia

### Smoking status

Our two groups of samples are: samples with COPD and samples with No
COPD. We calculate the means of the values of the expression levels.

```{r}
current <- rowMeans(logCPM[, se.filt$smoker=="Current smoker"])
former <- rowMeans(logCPM[, se.filt$smoker=="Former smoker"])
```

We can now calculate fold-change values for each gene in these two
groups. We'll rank the genes by decreasing absolute value of Fold
Change, and we'll call differentially expressed genes those that are at
the top 10.

```{r}
log2fc_smoker <- current - former
ranking <- order(abs(log2fc_smoker), decreasing=TRUE)
fold_change_smoker <- data.frame(EntrezID=names(log2fc_smoker[ranking]),
                                    Log2FC=round(log2fc_smoker[ranking], digits=3),
                                    FC=round(2^log2fc_smoker[ranking], digits=3),
                                    "1/FC"=round(2^(-log2fc_smoker[ranking]), digits=3),
                                    row.names=rowData(se.filt)$symbol[ranking],
                                    check.names=FALSE)
de_fc_smoker <- head(fold_change_smoker, n=10)
de_fc_smoker
```

In the resulting Data Frame:

-   If a gene has a **positive** Log2FC value, it means said gene is
    **FC** times more upregulated in samples that are *Current smokers*
    compared to in samples that are *Former smokers*.

    -   For example, AKR1B10 is [**26 times more
        upregulated**]{.underline} in samples that are Current smokers
        than in samples that are Former smokers

-   If a gene has a **negative** Log2FC value, it means said gene is
    **1/FC** times more upregulated in samples that are *Former smokers*
    compared to in samples that are *Current smokers*.

We can clearly see the difference of expression between groups is larger
with the smoking status distinction than the dysplasia one.

## Linear Regression to analyze Differential Expression

We will now use Linear Regression Models to perform our Differential
Expression, using the **limma pipeline**.

```{r}
library(limma)
```

We'll use the same two groupings, and use the followings steps:

1.  Build our design matrix
2.  Fit the linear model
3.  Calculate moderated *t*-statistics
4.  And finally we'll output the results, and get our list of
    Differentially Expressed Genes.

### Smoking status

We will start with smoking status since the fold-change values are
higher and it probably means we'll need less adjusting during the
process (and it'll be simpler).

First, we build our design matrix.

```{r}
mod <- model.matrix(~ smoker, data=colData(se.filt))
head(mod)
```

Next, we fit the linear regression model

```{r}
fit <- lmFit(assays(se.filt)$logCPM, mod)
```

We calculate moderated *t-*statistics

```{r}
fit <- eBayes(fit)
```

A quick overview of the results for FDR p-value cutoff at 5% (default)

```{r}
res <- decideTests(fit)
summary(res)

summary(res)[1,2] + summary(res)[3,2]
```

We have `r summary(res)[1,2]` downregulated genes, and
`r summary(res)[3,2]` upregulated genes, a total of
`r summary(res)[1,2] + summary(res)[3,2]`.

Seeing these results, we can afford to be a little more restricting with
the FDR cut-off. We'll fetch the results with a p-value cut-off of
0,001%.

```{r}
res <- decideTests(fit, p.value=0.00001)
summary(res)

summary(res)[1,2] + summary(res)[3,2]
```

We now have `r summary(res)[1,2]` downregulated genes, and
`r summary(res)[3,2]` upregulated genes, a total of
`r summary(res)[1,2] + summary(res)[3,2]`.

And a more detailed view of the results (we'll be sure to add more
information about the genes, so as to not only have the EntrezID
identifier as rowname).

The entire table, with information about all the genes:

```{r}
genesmd <- data.frame(chr=as.character(seqnames(rowRanges(se.filt))),
                      symbol=rowData(se.filt)[,5],
                      stringsAsFactors=FALSE)
fit$genes <- genesmd
  
tt_smoker <- topTable(fit, coef=2, n=Inf)
tt_smoker <- tt_smoker[order(tt_smoker$adj.P.Val),]
dim(tt_smoker)
head(tt_smoker)

```

But our goal is to get a list of DE genes, so, finally, we'll fetch the
genes called DE with FDR \<0,1%

```{r}
DEgenes_smoker <- rownames(tt_smoker)[tt_smoker$adj.P.Val < 0.00001]
length(DEgenes_smoker)
```

We'll now plot the distribution of *p*-values and moderated
*t*-statistics (Figure \@ref(fig:pdistSmoker))

```{r pdistSmoker, fig.cap="Distribution of p-values and moderated t-statistics on every gene between samples with that are Current smokers vs. samples that are former smokers."}
par(mfrow=c(1,2), mar=c(4, 5, 2, 2))
hist(tt_smoker$P.Value, xlab="Raw P-values", main="", las=1)
qqt(fit$t[, 2], df=fit$df.prior+fit$df.residual, main="", pch=".", cex=3)
abline(0, 1, lwd=2)
```

Another relevant one is the volcano plot

```{r}
volcanoplot(fit, coef=2, highlight=7, names=fit$genes$symbol, las=1)
```

We can see from the plots our results are highly relevant and
statistically significant.

### PML (Dysplasia)

Before we start, we have to exclude the metaplasia group, seeing as it
was excluded in the paper and we need two groups only, dysplasia vs. no
dysplasia.

```{r}
se.filt <- se.filt[,se.filt$dysplasia!="metaplasia"]
se.filt$dysplasia <- factor(se.filt$dysplasia, levels = c("dysplasia", "no dysplasia"))
dim(se.filt)
```

First, we build our design matrix.

```{r}
mod <- model.matrix(~ dysplasia, data=colData(se.filt))
head(mod)
```

Next, we fit the linear regression model

```{r}
fit <- lmFit(assays(se.filt)$logCPM, mod)
```

We calculate moderated *t-*statistics

```{r}
fit <- eBayes(fit)
```

A quick overview of the results for FDR p-value cutoff at 5% (default)

```{r}
res <- decideTests(fit)
summary(res)

summary(res)[1,2] + summary(res)[3,2]
```

We have `r summary(res)[1,2]` downregulated genes, and
`r summary(res)[3,2]` upregulated genes, a total of
`r summary(res)[1,2] + summary(res)[3,2]`.

Seeing these results, we can afford to be a little more restricting with
the FDR cut-off. We'll fetch the results with a p-value cut-off of 1%.

```{r}
res <- decideTests(fit, p.value=0.01)
summary(res)

summary(res)[1,2] + summary(res)[3,2]
```

We now have `r summary(res)[1,2]` downregulated genes, and
`r summary(res)[3,2]` upregulated genes, a total of
`r summary(res)[1,2] + summary(res)[3,2]`.

And a more detailed view of the results (we'll be sure to add more
information about the genes, so as to not only have the EntrezID
identifier as rowname).

The entire table, with information about all the genes:

```{r}
genesmd <- data.frame(chr=as.character(seqnames(rowRanges(se.filt))),
                      symbol=rowData(se.filt)[,5],
                      stringsAsFactors=FALSE)
fit$genes <- genesmd
  
tt_dysplasia <- topTable(fit, coef=2, n=Inf)
tt_dysplasia <- tt_dysplasia[order(tt_dysplasia$adj.P.Val),]
dim(tt_dysplasia)
head(tt_dysplasia)

```

But our goal is to get a list of DE genes, so, finally, we'll fetch the
genes called DE with FDR \<1%

```{r}
DEgenes_dysplasia <- rownames(tt_dysplasia)[tt_dysplasia$adj.P.Val < 0.01]
length(DEgenes_dysplasia)
```

We'll now plot the distribution of *p*-values and moderated
*t*-statistics (Figure \@ref(fig:pdistDysplasia))

```{r pdistDysplasia, fig.cap="Distribution of p-values and moderated t-statistics on every gene between samples with PML vs. without PML."}
par(mfrow=c(1,2), mar=c(4, 5, 2, 2))
hist(tt_dysplasia$P.Value, xlab="Raw P-values", main="", las=1)
qqt(fit$t[, 2], df=fit$df.prior+fit$df.residual, main="", pch=".", cex=3)
abline(0, 1, lwd=2)
```

Another relevant one is the volcano plot

```{r}
volcanoplot(fit, coef=2, highlight=7, names=fit$genes$symbol, las=1)
```

We can see from the plots our results are highly relevant and
statistically significant.

# Functional analysis

Now, let's attempt to find DE pathways, as DE gene sets, by detecting the functional enrichment. 

## Gene Ontology

Firstly, we will attempt to do it with the Gene Ontology (GO) analysis which corresponds to applying the
one-tailed Fisher's exact test to every gene set in the GO database. To execute this, we will use the
Bioconductor package GOstats.

The first step is to create our gene universe.

```{r, message=FALSE}
library(GOstats)
library(org.Hs.eg.db)

geneUniverse <- rownames(se.filt)
```


### Functional Analysis for Smoking Status

Let's build the parameter object and use the conditional test which computes the significance of a GO term conditional on the significance of its children.

```{r, message=FALSE}
paramsSmokerStatus <- new("GOHyperGParams", geneIds=DEgenes_smoker,    
                 universeGeneIds=geneUniverse,
                 annotation="org.Hs.eg.db", ontology="BP",
                 pvalueCutoff=0.05, testDirection="over")

conditional(paramsSmokerStatus) <- TRUE
```

The next step is to run the functional enrichment analysis.

```{r, message=FALSE}
hgOverCondSmokerStatus <- hyperGTest(paramsSmokerStatus)
hgOverCondSmokerStatus
```
And lastly to store and visualize the results. The resHgOverCond data.frame object contacts the results of the hypothesis test for each GO term,
satisfying the P-value cutoff.

```{r, message=FALSE}
resHgOverCondSmokerStatus <- summary(hgOverCondSmokerStatus)
dim(resHgOverCondSmokerStatus)
head(resHgOverCondSmokerStatus)

htmlReport(hgOverCondSmokerStatus, file="../doc/gotests_smoker_status.html")
browseURL("gotests_smoker_status.html")
```
Check out the result <a href="../doc/gotests_smoker_status.html" target="_blank">here</a>.

We do not want our GO terms to be involving too few genes or too many genes. That would make them too unreliable. So let's filter the previous results with a minimum size of 3, a maximum size of 300 and a minimum count of 3. Then we order them by the OddsRatio column.

```{r, message=FALSE}
resHgOverCondSmokerStatus <- resHgOverCondSmokerStatus[!is.infinite(resHgOverCondSmokerStatus$OddsRatio), ]
mask <- resHgOverCondSmokerStatus$Size >= 3 & resHgOverCondSmokerStatus$Size <= 300 & resHgOverCondSmokerStatus$Count >= 3
resHgOverCondSmokerStatus <- resHgOverCondSmokerStatus[mask, ]
dim(resHgOverCondSmokerStatus)

ord <- order(resHgOverCondSmokerStatus$OddsRatio, decreasing=TRUE)
resHgOverCondSmokerStatus <- resHgOverCondSmokerStatus[ord, ]
head(resHgOverCondSmokerStatus)
```

Using the geneIdsByCategory() method, we extract the genes that enrich each GO term and paste it to the result.

```{r, message=FALSE}
geneIDs <- geneIdsByCategory(hgOverCondSmokerStatus)[resHgOverCondSmokerStatus$GOBPID]
geneSYMs <- sapply(geneIDs, function(id) select(org.Hs.eg.db, columns="SYMBOL", key=id, keytype="ENTREZID")$SYMBOL)
geneSYMs <- sapply(geneSYMs, paste, collapse=", ")
goresultsSmokerStatus <- cbind(resHgOverCondSmokerStatus, Genes=geneSYMs)
rownames(goresultsSmokerStatus) <- 1:nrow(goresultsSmokerStatus)
```
We generate an HTML page using the knitr and kableExtra packages.

```{r GOTableSmokerStatusResults, echo=FALSE}
library(kableExtra)
ktab <- kable(goresultsSmokerStatus, "html", caption="GO results Smoker Status")
ktab <- kable_styling(ktab, bootstrap_options=c("stripped", "hover", "responsive"), fixed_thead=TRUE)
save_kable(ktab, file="../doc/goresults_smoker_status.html", self_contained=TRUE)
```

See the HTML page by clicking on this <a href="../doc/goresults_smoker_status.html" target="_blank">link</a>.

Table \@ref(tab:GOTableSmokerStatus2) displays the top 10 Go results (sorted by Odds Ratio).

```{r GOTableSmokerStatus2, echo=FALSE}
ktab <- kable(goresultsSmokerStatus[1:10, 2:7], "html", escape=FALSE, row.names=TRUE,
              caption=sprintf("GO Enrichment Smoker Status"))
kable_styling(ktab, position="center")
```

### Functional Analysis for PML (Dysplasia)

Build the parameter object and use the conditional test.

```{r, message=FALSE}
paramsDysplasia <- new("GOHyperGParams", geneIds=DEgenes_dysplasia,    
                 universeGeneIds=geneUniverse,
                 annotation="org.Hs.eg.db", ontology="BP",
                 pvalueCutoff=0.05, testDirection="over")

conditional(paramsDysplasia) <- TRUE
```

Run the functional enrichment analysis.

```{r, message=FALSE}
hgOverCondDysplasia <- hyperGTest(paramsDysplasia)
hgOverCondDysplasia
```

Store and visualize the results.

```{r, message=FALSE}
resHgOverCondDysplasia <- summary(hgOverCondDysplasia)
dim(resHgOverCondDysplasia)
head(resHgOverCondDysplasia)

htmlReport(hgOverCondDysplasia, file="../doc/gotests_dysplasia.html")
browseURL("gotests_dysplasia.html")
```

Check out the result <a href="../doc/gotests_dysplasia.html" target="_blank">here</a>.

Filter the previous results with a minimum size of 3, a maximum size of 300 and a minimum count of 3 and order them by the OddsRatio column.

```{r, message=FALSE}
resHgOverCondDysplasia <- resHgOverCondDysplasia[!is.infinite(resHgOverCondDysplasia$OddsRatio), ]
mask <- resHgOverCondDysplasia$Size >= 3 & resHgOverCondDysplasia$Size <= 300 & resHgOverCondDysplasia$Count >= 3
resHgOverCondDysplasia <- resHgOverCondDysplasia[mask, ]
dim(resHgOverCondDysplasia)

ord <- order(resHgOverCondDysplasia$OddsRatio, decreasing=TRUE)
resHgOverCondDysplasia <- resHgOverCondDysplasia[ord, ]
head(resHgOverCondDysplasia)
```

Using the geneIdsByCategory() method, we extract the genes that enrich each GO term and paste it to the result.

```{r, message=FALSE}
geneIDs <- geneIdsByCategory(hgOverCondDysplasia)[resHgOverCondDysplasia$GOBPID]
geneSYMs <- sapply(geneIDs, function(id) select(org.Hs.eg.db, columns="SYMBOL", key=id, keytype="ENTREZID")$SYMBOL)
geneSYMs <- sapply(geneSYMs, paste, collapse=", ")
goresultsDysplasia <- cbind(resHgOverCondDysplasia, Genes=geneSYMs)
rownames(goresultsDysplasia) <- 1:nrow(goresultsDysplasia)
```

We generate an HTML page using the knitr and kableExtra packages.

```{r GOTableDysplasiaResults, echo=FALSE}
library(kableExtra)
ktab <- kable(goresultsDysplasia, "html", caption="GO results Dysplasia")
ktab <- kable_styling(ktab, bootstrap_options=c("stripped", "hover", "responsive"), fixed_thead=TRUE)
save_kable(ktab, file="../doc/goresults_dysplasia.html", self_contained=TRUE)
```

See the HTML page by clicking on this <a href="../doc/goresults_dysplasia.html" target="_blank">link</a>.

Table \@ref(tab:GOTableDysplasia2) displays the top 10 Go results (sorted by Odds Ratio).

```{r GOTableDysplasia2, echo=FALSE}
ktab <- kable(goresultsDysplasia[1:10, 2:7], "html", escape=FALSE, row.names=TRUE,
              caption=sprintf("GO Enrichment Dysplasia"))
kable_styling(ktab, position="center")
```

## Gene Set Enrichment Analysis

The second part of the functional analysis involves performing Gene Set Enrichment Analysis (GSEA), a computational method used to determine if predefined sets of genes show statistically significant and concordant differences between two biological states (e.g., phenotypes). In our study, we will conduct GSEA for two comparisons: identifying pathways significantly enriched between the presence and absence of premalignant lesions (PMLs) and between current and former smokers.

### GSEA for Dysplasia (PML)

We load the necessary libraries: fgsea for GSEA, GSVAdata for gene set data, and GSEABase for gene set collection handling. We load the c2BroadSets dataset, which contains gene sets from KEGG, REACTOME, and BIOCARTA pathways. 
```{r}
library(fgsea)
library(GSVAdata)
library(GSEABase)
data(c2BroadSets)

c2BroadSets <- c2BroadSets[c(grep("^KEGG", names(c2BroadSets)),
  +                              grep("^REACTOME", names(c2BroadSets)),
  +                              grep("^BIOCARTA", names(c2BroadSets)))]
c2BroadSets

```

We create a custom gene set (de_fc_dysplasiaGS) using the Entrez IDs of differentially expressed genes in dysplasia. We then combine this custom gene set with the previously filtered gene sets into a single Gene Set Collection (gsc)

```{r}
de_fc_dysplasiaGS <- GeneSet(EntrezIdentifier("org.Hs.eg.db"), geneIds=de_fc_dysplasia$EntrezID, setName="DE_FC_DYS")
details(de_fc_dysplasiaGS)
gsc_dysplasia <- GeneSetCollection(c(c2BroadSets, de_fc_dysplasiaGS))
gsets_dysplasia <- geneIds(gsc_dysplasia)
```

We prepare the data by converting the histology information into factors and then into integers.

```{r}


histology_factors_dysplasia <- factor(se.filt$dysplasia)
histology_integers_dysplasia <- as.integer(histology_factors_dysplasia)

levels(histology_factors_dysplasia)
histology_integers_dysplasia
```

We use the fgseaLabel function to perform GSEA by permuting phenotypes 10,000 times to compute enrichment scores and p-values for each gene set.
```{r}

gseapermpheno_dysplasia <- fgseaLabel(gsets_dysplasia, assays(se.filt)$logCPM, histology_integers_dysplasia,10000, minSize=5, maxSize=300)
head(gseapermpheno_dysplasia[order(gseapermpheno_dysplasia$pval)])

```

We prepare a ranked list of genes based on their differential expression statistics (tt_dysplasia$t).
We then use the fgsea function to perform GSEA on this ranked list.

``` {r}
stats_dysplasia <- tt_dysplasia$t
names(stats_dysplasia) <- rownames(tt_dysplasia)
gseapermgsets_dysplasia <- fgsea(gsets, stats, minSize=5, maxSize=300)
head(gseapermgsets_dysplasia[order(gseapermgsets_dysplasia$pval), ])
gseapermpheno_dysplasia[gseapermpheno_dysplasia$padj < 0.01, ]
gseapermgsets_dysplasia[gseapermgsets_dysplasia$padj < 0.01, ]
```


We create histograms of p-values from both GSEA methods (permuting phenotypes and permuting gene sets) to visualize the distribution of p-values.
We also generate an enrichment plot for the custom gene set DE_FC_DYS to show the enrichment score across the ranked gene list.

```{r}
par(mfrow=c(1, 2))
hist(gseapermpheno_dysplasia$pval, xlab="P-value", main="Permuting phenotypes", las=1)
hist(gseapermgsets_dysplasia$pval, xlab="P-value", main="Permuting gene sets", las=1)
plotEnrichment(gsets_dysplasia$DE_FC_DYS, stats_dysplasia)

```

We use this part of the code to identify pathways that are significantly enriched (adjusted p-value < 0.01). The plotGseaTable function then generates a table and plot of GSEA results for these significant pathways, showing the enrichment scores and p-values. This helps us visualize and interpret the most relevant pathways in our dysplasia analysis.

```{r}
DEpwys_dysplasia <- gseapermgsets_dysplasia$pathway[gseapermgsets_dysplasia$padj < 0.01]
plotGseaTable(gsets_dysplasia[DEpwys_dysplasia], stats_dysplasia, gseapermgsets_dysplasia, gseaParam=0.5)

```

We use the limma::camera function to perform an additional gene set analysis that adjusts for inter-gene correlation within gene sets. This step enhances the robustness of our findings by providing an alternative method to validate the significant pathways identified by GSEA. The camera function uses the same gene sets and design matrix as our previous analyses.

```{r} 
library(limma)
gsetidx_dysplasia <- ids2indices(geneIds(gsc_dysplasia), rownames(se.filt))
camerares_dysplasia <- camera(y=assays(se.filt)$logCPM, index=gsetidx_dysplasia, design=mod, contrast=2)
head(camerares_dysplasia, 5)
```

We filter the Camera results to focus on pathways with an FDR < 0.01, ensuring we only consider the most significant pathways. This step helps us further refine our list of relevant pathways.

```{r}
camerares_dysplasia[camerares_dysplasia$FDR < 0.01, ]
```

We create a histogram of the p-values from the Camera results to visualize the distribution of significance levels across all tested pathways. This plot provides an overview of the significance of the pathways in relation to dysplasia.

```{r}
hist(camerares_dysplasia$PValue, xlab="P-value", main="Camera", las=1)
```


### GSEA for Smoking status 

We do the same steps as previously but now for the smoking status.
Load the required libraries and data.

```{r}
library(fgsea)
library(GSVAdata)
library(GSEABase)
data(c2BroadSets)

c2BroadSets <- c2BroadSets[c(grep("^KEGG", names(c2BroadSets)),
  +                              grep("^REACTOME", names(c2BroadSets)),
  +                              grep("^BIOCARTA", names(c2BroadSets)))]
c2BroadSets
```

Create a costum gene set for Smoking status.

```{r}
de_fc_smokerGS <- GeneSet(EntrezIdentifier("org.Hs.eg.db"), geneIds=de_fc_smoker$EntrezID, setName="DE_FC_SMOKER")
details(de_fc_smokerGS)
gsc_smoker <- GeneSetCollection(c(c2BroadSets, de_fc_smokerGS))
gsets_smoker <- geneIds(gsc_smoker)
```

Prepare the data. 

```{r}
histology_factors_smoker <- factor(se.filt$smoker)
histology_integers_smoker <- as.integer(histology_factors_smoker)

levels(histology_factors_smoker)
histology_integers_smoker

```

Run the GSEA by permuting phenotypes.
```{r}
gseapermpheno_smoker <- fgseaLabel(gsets_smoker, assays(se.filt)$logCPM, histology_integers_smoker,10000, minSize=5, maxSize=300)
head(gseapermpheno_smoker[order(gseapermpheno_smoker$pval), ])
```

Run GSEA with the ranked gene list.
``` {r}
stats_smoker <- tt_smoker$t
names(stats_smoker) <- rownames(tt_smoker)
gseapermgsets_smoker <- fgsea(gsets_smoker, stats_smoker, minSize=5, maxSize=300)
head(gseapermgsets_smoker[order(gseapermgsets_smoker$pval), ])
gseapermpheno_smoker[gseapermpheno_smoker$padj < 0.01, ]
gseapermgsets_smoker[gseapermgsets_smoker$padj < 0.01, ]
```

Visualize GSEA results.
```{r}
par(mfrow=c(1, 2))
hist(gseapermpheno_smoker$pval, xlab="P-value", main="Permuting phenotypes", las=1)
hist(gseapermgsets_smoker$pval, xlab="P-value", main="Permuting gene sets", las=1)
plotEnrichment(gsets_smoker$DE_FC_SMOKER, stats_smoker)

```

Identify significant pathways.

```{r}
DEpwys <- gseapermgsets_smoker$pathway[gseapermgsets_smoker$padj < 0.01]
plotGseaTable(gsets_smoker[DEpwys], stats_smoker, gseapermgsets_smoker, gseaParam=0.5)

```

Additional validation using Camera.

```{r} 
library(limma)
gsetidx_smoker <- ids2indices(geneIds(gsc_smoker), rownames(se.filt))
camerares_smoker <- camera(y=assays(se.filt)$logCPM, index=gsetidx_smoker, design=mod, contrast=2)
head(camerares_smoker, 5)
```

Filter significant pathways from Camera.

```{r}
camerares_smoker[camerares_smoker$FDR < 0.01, ]
```

Visualise significant pathways from Camera for Smoking status.
```{r}
hist(camerares_smoker$PValue, xlab="P-value", main="Camera", las=1)
```

# Discussion

## Current vs Former Smoker Results

### Go Ontology

The Go terms with the highest Odds Ratio were tryptophan transport, valine transport and peptidyl-aspartic acid modification. This suggests that these processes are significantly different between current and former smokers.

Tryptophan Transport (GO:0015827): Tryptophan is a precursor to serotonin, a neurotransmitter involved in mood regulation. Smoking has been associated with altered serotonin levels and mood disorders.
Current smokers might have different tryptophan transport dynamics due to chronic exposure to nicotine, which could affect neurotransmitter synthesis and mental health, whereas former smokers might show a return to baseline levels.

Valine Transport (GO:0015829): Valine is a branched-chain amino acid (BCAA) important for muscle metabolism and energy production. Smoking affects metabolic processes and energy balance.
The differences in valine transport could reflect altered energy metabolism in current smokers, who may experience metabolic stress due to smoking-related oxidative damage.

Peptidyl-Aspartic Acid Modification (GO:0018197): Post-translational modifications of proteins are crucial for their function. Smoking can induce oxidative stress, leading to changes in protein modification.
Current smokers might have altered post-translational modifications due to the high oxidative stress from smoking, affecting protein function and cellular processes.

Terpenoid Catabolic Process (GO:0016115): When it comes to Terpenoid Catabolic Process (GO:0016115), terpenoids include many bioactive compounds, some of which are found in tobacco smoke. The metabolism of these compounds is critical for detoxification. Current smokers may exhibit increased catabolism of terpenoids as their bodies work to break down harmful components of tobacco smoke, whereas former smokers may show reduced activity as the exposure decreases.

Daunorubicin Metabolic Process (GO:0044597) and Doxorubicin Metabolic Process (GO:0044598): Daunorubicin and doxorubicin are metabolized in the body and smoking is known to influence drug metabolism. The differences observed could indicate that current smokers have altered drug metabolism capabilities, which might affect the efficacy and toxicity of medications compared to former smokers. 

Sinoatrial Node Development (GO:0003163): The sinoatrial node is critical for heart rhythm regulation, and smoking is a known risk factor for cardiovascular diseases.
Current smokers might exhibit changes in the development or function of the sinoatrial node, possibly due to chronic cardiovascular stress and exposure to harmful substances in smoke.

Taurine Transport (GO:0015734): Taurine plays a role in cellular osmoregulation, bile salt formation, and antioxidation. Smoking affects many of these physiological processes. Differences in taurine transport could reflect adaptations to oxidative stress and altered bile acid metabolism in current smokers compared to former smokers.

Isoleucine Transport (GO:0015820): Isoleucine, like valine, is a BCAA important for protein synthesis and energy metabolism. Smoking impacts overall metabolic health.
Altered transport of isoleucine in current smokers might be a response to the increased metabolic demands and stress associated with chronic smoking.

Thyroid Hormone Transport (GO:0070327): Thyroid hormones regulate metabolism, and smoking has been linked to changes in thyroid function.
The differences in thyroid hormone transport could indicate that smoking affects the regulation and distribution of these hormones, impacting metabolic processes and overall health.

## Dysplasia vs No Dysplasia Results

7-methylguanosine cap hypermethylation (GO:0036261): This process is related to the modification of the mRNA cap structure, which is crucial for mRNA stability and translation. Smoking can induce changes in RNA processing and stability, potentially through increased oxidative stress and inflammation. These changes could lead to dysregulation of gene expression, contributing to dysplasia.

Negative regulation of protein neddylation (GO:2000435): Neddylation is a post-translational modification affecting protein stability and function.
Dysregulation of protein modifications, including neddylation, can be influenced by the cellular stress and inflammatory response induced by smoking. This could lead to altered protein function and contribute to the pathogenesis of dysplasia.

Cytoplasmic translation (GO:0002181): Involves the synthesis of proteins in the cytoplasm.Smoking can affect protein synthesis by inducing endoplasmic reticulum stress and disrupting ribosomal function. Changes in protein synthesis are often seen in cancer and precancerous conditions like dysplasia.

Mitochondrial electron transport, cytochrome c to oxygen (GO:0006123) & ubiquinol to cytochrome c (GO:0006122): Part of the mitochondrial respiratory chain involved in energy production. Smoking increases oxidative stress, which can damage mitochondria and impair electron transport. Mitochondrial dysfunction is a known factor in the development of many diseases, including cancer and dysplasia.

Positive regulation of miRNA catabolic process (GO:2000627): miRNAs are small non-coding RNAs that regulate gene expression post-transcriptionally. Smoking can alter miRNA expression profiles, contributing to changes in gene regulation that can lead to dysplasia and cancer.

Proton motive force-driven ATP synthesis (GO:0015986): This process is a part of cellular respiration where ATP is produced. Mitochondrial dysfunction due to smoking can disrupt ATP synthesis, leading to cellular energy deficits and promoting disease processes like dysplasia.

Protein maturation by [4Fe-4S] cluster transfer (GO:0106035): This involves the assembly of iron-sulfur clusters, which are essential for various cellular processes, including DNA repair and electron transport. Oxidative stress from smoking can impair iron-sulfur cluster assembly, leading to defects in cellular metabolism and potentially contributing to dysplasia.

Ribosomal small subunit assembly (GO:0000028) & Ribosomal small subunit biogenesis (GO:0042274): These are critical for the formation of functional ribosomes, which are essential for protein synthesis. Disruption in ribosome biogenesis can be caused by smoking-related oxidative stress and inflammation, affecting protein synthesis and cell growth, potentially leading to dysplasia.


## GSEA

### Dysplasia vs no dysplasia results 

The most significant pathways identified in this analysis are primarily related to protein synthesis and ribosome function, indicating that these processes are notably dysregulated in dysplasia. The high significance of these pathways (with p-values less than 1e-50) underscores their potential role in the pathogenesis of dysplasia and their importance as areas for further investigation. The pathways are listed below: 

KEGG_RIBOSOME
This pathway involves the synthesis and function of ribosomes, essential for protein synthesis. Smoking can disrupt ribosomal biogenesis and function, leading to altered protein synthesis. Such disruptions are common in rapidly dividing cancer cells and precancerous conditions like dysplasia, where proper protein synthesis is vital for normal cell growth and differentiation.

REACTOME_PEPTIDE_CHAIN_ELONGATION
Peptide chain elongation is a critical step in translation where amino acids are added to the growing peptide chain. Smoking can lead to dysregulated elongation through oxidative stress and inflammatory responses, resulting in abnormal protein synthesis and contributing to cellular transformation seen in dysplasia and cancer.

REACTOME_VIRAL_MRNA_TRANSLATION
This pathway involves the translation of viral mRNA within host cells. Smoking-induced inflammation and oxidative stress can create a cellular environment that facilitates viral infections and disrupts normal mRNA translation processes, contributing to cellular dysfunction and the development of dysplasia and cancer.

REACTOME_FORMATION_OF_A_POOL_OF_FREE_40S_SUBUNITS
This pathway involves the formation of free 40S ribosomal subunits, which are crucial for the initiation of translation. Smoking can disrupt this process through oxidative stress and ribosomal damage, leading to impaired protein synthesis. Such dysregulation can contribute to uncontrolled cell growth and the development of dysplasia and cancer.

REACTOME_INFLUENZA_VIRAL_RNA_TRANSCRIPTION_AND_REPLICATION
This pathway describes the transcription and replication of influenza viral RNA, which can disrupt host cellular processes. Smoking-induced inflammation and oxidative stress can create an environment conducive to viral infections, further disrupting normal cellular functions and contributing to the development of dysplasia and cancer.

### Current and former smoker results

BIOCARTA_IL7_PATHWAY:
The IL-7 pathway is essential for normal development of B cells and T cells in the immune system.
Mice lacking the IL-7 receptor exhibit a deficiency in both B and T cells. Some humans with severe combined immunodeficiency disease (SCID) also have mutations in their IL-7 receptor gene, leading to impaired B cell production and T cell absence.
For current smokers, chronic nicotine exposure may impact IL-7 signaling, affecting immune responses.
Former smokers might see a return to baseline IL-7 signaling dynamics after quitting.

BIOCARTA_PROTEASOME_PATHWAY:
This pathway involves the proteasome, responsible for protein degradation.
Proteasomes regulate protein turnover, impacting various cellular processes.
Current smokers: Smoking-related oxidative stress may affect proteasome function.
Former smokers: Proteasome activity may normalize after quitting.

BIOCARTA_SALMONELLA_PATHWAY:
Relates to Salmonella infection.
Salmonella is a pathogenic bacterium causing foodborne illness.
Current smokers: Altered immune responses due to smoking might impact susceptibility to infections like Salmonella.
Former smokers: Reduced exposure may decrease susceptibility.

KEGG_ALLOGRAFT_REJECTION:
Involves immune responses against transplanted tissues (allografts).
Allograft rejection is influenced by immune activation.
Current smokers: Smoking-related immune changes could affect allograft acceptance or rejection.
Former smokers: Reduced exposure may improve allograft outcomes.

KEGG_ALZHEIMERS_DISEASE:
Relates to Alzheimer’s disease.
Alzheimer’s involves neuroinflammation and protein misfolding.
Current smokers: Smoking-associated oxidative stress and inflammation may increase Alzheimer’s risk.
Former smokers: Risk reduction after quitting.


## Results

In our project, we aimed to explore the effects of smoking on cellular functions using Gene Set Enrichment Analysis (GSEA) and Gene Ontology (GO) analysis. We conducted these analyses on samples with dysplasia versus no dysplasia and current smokers versus former smokers.

Both GSEA and GO analysis revealed that smoking significantly impacts various cellular pathways. A key finding across both analyses was the consistent identification of pathways related to ribosomal function and protein synthesis. Specifically, we found that smoking affects the expression and function of ribosomal proteins, which are critical for protein biogenesis.

Our findings align with the study by Park et al. (2021) published in Scientific Reports, which also identified a detrimental effect of smoking on ribosomal function and protein synthesis. This study found that exposure to electronic cigarette smoke reduces ribosomal protein gene expression, impairing protein synthesis in primary human airway epithelial cells. The reduced expression of ribosomal proteins and subsequent impairment in protein synthesis could lead to various cellular dysfunctions.

The impairment of ribosomal function may contribute to the pathogenesis of various smoking-related diseases, including cancer. The referenced study highlighted the potential for smoking to induce carcinogenesis through the disruption of ribosomal biogenesis and protein synthesis, which are critical for maintaining cellular homeostasis.

Further findings indicate that smoking disrupts cytoplasmic translation and peptide chain elongation.The adverse effects of smoking on ribosomal function and protein synthesis were further supported by the study of Merzah et al. (2023), which identified differentially expressed genes (DEGs) related to immune response and inflammation pathways in smokers. The study revealed that smoking leads to a significant downregulation of genes involved in oxygen binding and activity, emphasizing the role of oxidative stress in disrupting cellular functions.

The findings highlight the potential role of impaired protein synthesis in the pathogenesis of smoking-related diseases, including cancer and coronary artery disease (CAD). The study by Merzah et al. (2023) supports the link between smoking, oxidative stress, and the disruption of critical cellular pathways, contributing to disease development.

Other pathways identified in our analyses were related to mitochondrial electron transport and ATP synthesis, emphasizing how smoking-induced oxidative stress disrupts cellular functions. Smoking introduces numerous toxins that induce oxidative stress, resulting in the production of reactive oxygen species (ROS) which can damage cellular components, including mitochondria. This oxidative stress impairs mitochondrial function, leading to a cascade of detrimental effects on cellular health. As highlighted in the study by Kanithi et al. (2022), toxins in cigarette smoke and e-cigarette vapor significantly damage mitochondrial function by increasing oxidative stress, disrupting mitochondrial morphology, and impairing critical processes such as fusion, fission, and mitophagy. 

Our analyses also identified significant disruptions in pathways related ATP synthesis. These processes are crucial for cellular energy production and overall mitochondrial health. Smoking-induced oxidative stress damages the electron transport chain, leading to reduced ATP production and impaired cellular energy metabolism. The resulting energy deficit can hinder cellular processes, contributing to the development of diseases such as cancer and cardiovascular conditions.



# Conclusions

Here we summarize our conclusions.

# Session information

```{r}
sessionInfo()
```

# References
